ZivyObraz server API

I like how fw was updating recently and I think it might be time for this.

Since there are more sensors and possibly there will be others, I propose that the server API uses http body instead of url parameters. I know, it makes testing a bit more difficult but the benefits may be worth it.

The API shall use JSON in body. It is flexible and easy to parse. This way we could add more not only sensors but any data we want. We could put the API key here, etc...

What do you think of that? I can Help implementing it.

Here is the first draft of how it could look like. What do you think of that?

API key/PIN I would put into request header. As far as I remember it is standard. Mainly because GET requests usually do not contain body.

This would be a POST request

```json
{
    "apiVersion": "3.0",
    "board": "ESPink-v3",
    "fwVersion": "LaskaKit-0.4",
    "timestampCheck": true,
    "system": {
        "chip": "esp32s3"
        "uptime": 7878,
        "cpuTemp": 44.3,
        "resetReason": "deepsleep"
    },
    "network": {
        "ssid": "ssid",
        "rssi": -60,
        "mac": "macaddr",
        "apRetries": 0,
        "lastDownloadDuration": 5
    },
    "display": {
        "type": "GDEMxxx",
        "width": 800,
        "height": 480,
        "colorType": "4C",
        "lastRefreshDuration": 23
    },
    "sensors": [
        {
            "type": "voltage",
            "v": 3.5
        },
        {
            "type": "sht40",
            "temp": 23.4,
            "hum": 60.2
        },
        {
            "type": "bme280",
            "temp": 22.3,
            "hum": 59.2,
            "pres": 1000
        }
    ],
    "customValues": [
        {
            "key": "strValue",
            "value": "str",
            "tag": "customTag(optional)"
        },
        {
            "key": "someIntValue",
            "value": 123,
            "tag": "customTag(optional)"
        },
        {
            "key": "someArray",
            "value": [1, 2, 3, 4, 5],
            "tag": "customTag(optional)"
        }
    ],
    "backlog": [
        "[ERROR] Decoder - overflow",
        "[ERROR] Display - buffer allocation failed"
    ]
}
```

--- a/README.md
+++ b/README.md
@@ -17,6 +17,7 @@ You will need to have the following libraries installed:
 ```ini
 lib_deps =
     zinggjm/GxEPD2@^1.6.0
+    bblanchon/ArduinoJson@^7.2.1
     adafruit/Adafruit GFX Library@^1.11.9
     adafruit/Adafruit SHT4x Library@^1.0.5
     adafruit/Adafruit BME280 Library@^2.3.0

--- a/platformio.ini
+++ b/platformio.ini
@@ -19,6 +19,7 @@ lib_deps_builtin =
     Wire
 lib_deps =
     zinggjm/GxEPD2@^1.6.5
+    bblanchon/ArduinoJson@^7.2.1
     adafruit/Adafruit GFX Library@^1.11.9
     adafruit/Adafruit SHT4x Library@^1.0.5
     adafruit/Adafruit BME280 Library@^2.3.0

--- a/src/board.cpp
+++ b/src/board.cpp
@@ -94,7 +94,7 @@ float getBatteryVoltage()
   float volt;
 
 #ifdef ESPink_V3
-  Serial.println("[Battery] Readingon ESPink V3 board");
+  Serial.println("[BATTERY] Readingon ESPink V3 board");
 
   setEPaperPowerOn(true);
   pinMode(PIN_ALERT, INPUT_PULLUP);
@@ -108,13 +108,13 @@ float getBatteryVoltage()
   // lipo.enableDebugging(); // Uncomment this line to enable helpful debug messages on Serial
 
   // Read and print the reset indicator
-  Serial.print(F("[Battery] Reset Indicator was: "));
+  Serial.print(F("[BATTERY] Reset Indicator was: "));
   bool RI = lipo.isReset(true); // Read the RI flag and clear it automatically if it is set
   Serial.println(RI);           // Print the RI
   // If RI was set, check it is now clear
   if (RI)
   {
-    Serial.print(F("[Battery] Reset Indicator is now: "));
+    Serial.print(F("[BATTERY] Reset Indicator is now: "));
     RI = lipo.isReset(); // Read the RI flag
     Serial.println(RI);  // Print the RI
   }
@@ -126,7 +126,7 @@ float getBatteryVoltage()
   volt = (float)lipo.getVoltage();
   // percentage could be read like this:
   // lipo.getSOC();
-  // Serial.println("[Battery] Percentage: " + String(lipo.getSOC(), 2) + " %");
+  // Serial.println("[BATTERY] Percentage: " + String(lipo.getSOC(), 2) + " %");
 
   lipo.clearAlert();
   lipo.enableHibernate();
@@ -138,7 +138,7 @@ float getBatteryVoltage()
   esp_adc_cal_characterize(ADC_UNIT_1, ADC_ATTEN_DB_12, ADC_WIDTH_BIT_12, 0, &adc_cal);
   adc1_config_channel_atten(vBatPin, ADC_ATTEN_DB_12);
 
-  Serial.println("[Battery] Reading on ES3ink board");
+  Serial.println("[BATTERY] Reading on ES3ink board");
 
   digitalWrite(enableBattery, LOW);
   uint32_t raw = adc1_get_raw(vBatPin);
@@ -151,7 +151,7 @@ float getBatteryVoltage()
   digitalWrite(enableBattery, HIGH);
 
 #elif defined ESP32S3Adapter
-  Serial.println("[Battery] Reading on ESP32-S3 DEVKIT board");
+  Serial.println("[BATTERY] Reading on ESP32-S3 DEVKIT board");
   // attach ADC input
   volt = (analogReadMilliVolts(vBatPin) * dividerRatio / 1000);
 
@@ -225,7 +225,7 @@ float getBatteryVoltage()
   pinMode(enableBattery, INPUT);
 
 #elif defined SEEEDSTUDIO_RETERMINAL
-  Serial.println("[Battery] Reading on SeeedStudio reTerminal board");
+  Serial.println("[BATTERY] Reading on SeeedStudio reTerminal board");
   // Enable battery voltage measurement circuit via GPIO21
   digitalWrite(enableBattery, HIGH);
   pinMode(enableBattery, OUTPUT);
@@ -238,7 +238,7 @@ float getBatteryVoltage()
   volt = (analogReadMilliVolts(vBatPin) * dividerRatio / 1000);
 #endif
 
-  Serial.println("[Battery] Voltage: " + String(volt) + " V");
+  Serial.println("[BATTERY] Voltage: " + String(volt) + " V");
   return volt;
 }
 
@@ -251,7 +251,7 @@ unsigned long checkButtonPressDuration()
   if (digitalRead(EXT_BUTTON) == HIGH)
     return 0; // Button not pressed
 
-  Serial.println("[Button] Press detected at boot, measuring duration...");
+  Serial.println("[BUTTON] Press detected at boot, measuring duration...");
 
   unsigned long pressStart = millis();
   const unsigned long maxWaitTime = 10000;
@@ -266,11 +266,81 @@ unsigned long checkButtonPressDuration()
   }
 
   unsigned long pressDuration = millis() - pressStart;
-  Serial.println("[Button] Press duration: " + String(pressDuration) + " ms");
+  Serial.println("[BUTTON] Press duration: " + String(pressDuration) + " ms");
   return pressDuration;
 #else
   // EXT_BUTTON not defined for this board
   return 0;
 #endif
+}
+
+float getCPUTemperature()
+{
+#if CONFIG_IDF_TARGET_ESP32 || CONFIG_IDF_TARGET_ESP32S2 || CONFIG_IDF_TARGET_ESP32S3 || CONFIG_IDF_TARGET_ESP32C3
+  return temperatureRead();
+#else
+  return 0.0f;
+#endif
+}
+
+ResetReason getResetReason()
+{
+  esp_reset_reason_t reason = esp_reset_reason();
+
+  switch (reason)
+  {
+    case ESP_RST_POWERON:
+      return ResetReason::POWERON;
+    case ESP_RST_EXT:
+      return ResetReason::EXT;
+    case ESP_RST_SW:
+      return ResetReason::SW;
+    case ESP_RST_PANIC:
+      return ResetReason::PANIC;
+    case ESP_RST_INT_WDT:
+      return ResetReason::INT_WDT;
+    case ESP_RST_TASK_WDT:
+      return ResetReason::TASK_WDT;
+    case ESP_RST_WDT:
+      return ResetReason::WDT;
+    case ESP_RST_DEEPSLEEP:
+      return ResetReason::DEEPSLEEP;
+    case ESP_RST_BROWNOUT:
+      return ResetReason::BROWNOUT;
+    case ESP_RST_SDIO:
+      return ResetReason::SDIO;
+    default:
+      return ResetReason::UNKNOWN;
+  }
+}
+
+const char *getResetReasonString()
+{
+  esp_reset_reason_t reason = esp_reset_reason();
+  switch (reason)
+  {
+    case ESP_RST_POWERON:
+      return "poweron";
+    case ESP_RST_EXT:
+      return "external";
+    case ESP_RST_SW:
+      return "software";
+    case ESP_RST_PANIC:
+      return "panic";
+    case ESP_RST_INT_WDT:
+      return "int_watchdog";
+    case ESP_RST_TASK_WDT:
+      return "task_watchdog";
+    case ESP_RST_WDT:
+      return "watchdog";
+    case ESP_RST_DEEPSLEEP:
+      return "deepsleep";
+    case ESP_RST_BROWNOUT:
+      return "brownout";
+    case ESP_RST_SDIO:
+      return "sdio";
+    default:
+      return "unknown";
+  }
 }
 } // namespace Board

--- a/src/board.h
+++ b/src/board.h
@@ -316,6 +316,21 @@ static constexpr const char BOARD_TYPE_STRING[] = XSTR(BOARD_TYPE);
   #define dividerRatio (1.769f)
 #endif
 
+enum class ResetReason : uint8_t
+{
+  UNKNOWN = 0,
+  POWERON = 1,   // Power-on reset or flash upload
+  EXT = 2,       // External reset (reset button)
+  SW = 3,        // Software reset
+  PANIC = 4,     // Software panic/exception
+  INT_WDT = 5,   // Interrupt watchdog
+  TASK_WDT = 6,  // Task watchdog
+  WDT = 7,       // Other watchdog
+  DEEPSLEEP = 8, // Wake from deep sleep
+  BROWNOUT = 9,  // Brownout reset
+  SDIO = 10      // Reset over SDIO
+};
+
 namespace Board
 {
 void setupHW();
@@ -323,6 +338,9 @@ void setEPaperPowerOn(bool on);
 void enterDeepSleepMode(uint64_t sleepDuration);
 
 float getBatteryVoltage();
+float getCPUTemperature();
+ResetReason getResetReason();
+const char *getResetReasonString();
 
 unsigned long checkButtonPressDuration();
 

--- a/src/display.cpp
+++ b/src/display.cpp
@@ -358,7 +358,7 @@ void init()
 
 void clear()
 {
-  Serial.println("[Display] Clearing...");
+  Serial.println("[DISPLAY] Clearing...");
 
   init();
 
@@ -377,7 +377,7 @@ void clear()
   // Disable power supply for ePaper
   Board::setEPaperPowerOn(false);
 
-  Serial.println("[Display] Clearing done");
+  Serial.println("[DISPLAY] Clearing done");
 }
 
 void setRotation(uint8_t rotation) { display.setRotation(rotation); }
@@ -497,7 +497,7 @@ void enableLightSleepDuringRefresh(bool enable)
 #ifndef M5StackCoreInk
   if (enable)
   {
-    Serial.println("[Sleep] Light sleep during display refresh");
+    Serial.println("[SLEEP] Light sleep during display refresh");
     display.epd2.setBusyCallback(busyCallbackLightSleep, nullptr);
   }
   else

--- a/src/http_client.cpp
+++ b/src/http_client.cpp
@@ -1,6 +1,9 @@
 #include "http_client.h"
 
+#include <ArduinoJson.h>
+
 #include "board.h"
+#include "sensor.h"
 #include "display.h"
 #include "state_manager.h"
 #include "utils.h"
@@ -15,35 +18,66 @@ HttpClient::HttpClient()
       m_serverTimestamp(0),
       m_displayRotation(0),
       m_hasRotation(false),
-      m_partialRefresh(false)
+      m_partialRefresh(false),
+      m_jsonPayload("")
 {
 }
 
-bool HttpClient::sendRequest(bool timestampCheckOnly, const String &extraParams)
+void HttpClient::buildJsonPayload()
 {
-  // Build URL
-  String url = "/index.php?mac=" + Wireless::getMacAddress();
+  // Only build once - if already built, skip
+  if (m_jsonPayload.length() > 0)
+    return;
 
-  if (timestampCheckOnly)
-    url += "&timestamp_check=1";
+  JsonDocument doc;
 
-  url += "&b=" + String(Board::getBoardType());
-  url += "&d=" + String(Display::getDisplayType());
-  url += "&rssi=" + String(Wireless::getStrength());
-  url += "&ssid=" + Wireless::getSSID();
-  url += "&v=" + String(Board::getBatteryVoltage());
-  url += "&x=" + String(Display::getResolutionX());
-  url += "&y=" + String(Display::getResolutionY());
-  url += "&c=" + String(Display::getColorType());
-  url += "&fw=" + String(firmware);
-  url += "&ap_retries=" + String(StateManager::getFailureCount());
-  url += "&r=" + String(static_cast<uint8_t>(StateManager::getResetReason()));
+  // API and firmware info
+  doc["fwVersion"] = firmware;
+  doc["apiVersion"] = firmware; // tells server what are firmware capabilities, same as fwVersion in our case
+  doc["board"] = Board::getBoardType();
 
-  // Add last compensation time if available from previous run
-  if (StateManager::getLastCompensationTime() > 0)
-    url += "&ct=" + String(StateManager::getLastCompensationTime());
+  // System info
+  JsonObject system = doc["system"].to<JsonObject>();
+  system["cpuTemp"] = Board::getCPUTemperature();
+  system["resetReason"] = Board::getResetReasonString();
+  system["vccVoltage"] = Board::getBatteryVoltage();
 
-  url += extraParams;
+  // Network info
+  JsonObject network = doc["network"].to<JsonObject>();
+  network["ssid"] = WiFi.SSID();
+  network["rssi"] = Wireless::getStrength();
+  network["mac"] = Wireless::getMacAddress();
+  network["apRetries"] = StateManager::getFailureCount();
+  network["ipAddress"] = Wireless::getIPAddress();
+  // Add last download duration if available from previous run (in milliseconds)
+  if (StateManager::getLastDownloadDuration() > 0)
+    network["lastDownloadDuration"] = StateManager::getLastDownloadDuration();
+
+  // Display info
+  JsonObject display = doc["display"].to<JsonObject>();
+  display["type"] = Display::getDisplayType();
+  display["width"] = Display::getResolutionX();
+  display["height"] = Display::getResolutionY();
+  display["colorType"] = Display::getColorType();
+  // Add last refresh duration if available from previous run (in milliseconds)
+  if (StateManager::getLastRefreshDuration() > 0)
+    display["lastRefreshDuration"] = StateManager::getLastRefreshDuration();
+
+  // Add sensor data if available
+  Sensor::SensorData sensorData = Sensor::getSensorData();
+  if (sensorData.isValid)
+  {
+    JsonArray sensors = doc["sensors"].to<JsonArray>();
+    sensorData.toJson(sensors);
+  }
+
+  serializeJson(doc, m_jsonPayload);
+}
+
+bool HttpClient::sendRequest(bool timestampCheck)
+{
+  // Build JSON payload (only built once, cached for subsequent requests)
+  buildJsonPayload();
 
   Serial.print("[HTTP] Connecting to: ");
   Serial.println(host);
@@ -64,22 +98,28 @@ bool HttpClient::sendRequest(bool timestampCheckOnly, const String &extraParams)
     if (attempt == 2)
     {
       m_sleepDuration = StateManager::DEFAULT_SLEEP_SECONDS;
-      if (!timestampCheckOnly)
+      if (!timestampCheck)
         return false; // For image download, fail immediately
       delay(500);
     }
-    if (!timestampCheckOnly)
+    if (!timestampCheck)
       delay(200);
   }
 
-  // Send HTTP/HTTPS request
-  Serial.print("[HTTP] Requesting URL: ");
+  // Send HTTP/HTTPS POST request with JSON body
+  Serial.print("[HTTP] Sending POST to: ");
   Serial.print(CONNECTION_URL_PREFIX);
   Serial.print(host);
-  Serial.println(url);
+  Serial.println("/index.php");
+  Serial.println("[HTTP] JSON payload: " + m_jsonPayload);
+  // Build URL with timestampCheck query parameter
+  String url = "/index.php?timestampCheck=";
+  url += timestampCheck ? "1" : "0";
 
-  m_client.print(String("GET ") + url + " HTTP/1.1\r\n" + "Host: " + host + "\r\n" +
-                 "X-API-Key: " + String(Utils::getStoredAPIKey()) + "\r\n" + "Connection: close\r\n\r\n");
+  m_client.print(String("POST ") + url + " HTTP/1.1\r\n" + "Host: " + host + "\r\n" +
+                 "X-API-Key: " + String(Utils::getStoredAPIKey()) + "\r\n" + "Content-Type: application/json\r\n" +
+                 "Content-Length: " + String(m_jsonPayload.length()) + "\r\n" + "Connection: close\r\n\r\n" +
+                 m_jsonPayload);
 
   Serial.println("[HTTP] Request sent");
 
@@ -91,7 +131,7 @@ bool HttpClient::sendRequest(bool timestampCheckOnly, const String &extraParams)
     {
       Serial.println("[HTTP] >>> Client Timeout!");
       m_client.stop();
-      if (timestampCheckOnly)
+      if (timestampCheck)
         m_sleepDuration = StateManager::DEFAULT_SLEEP_SECONDS;
       return false;
     }
@@ -137,11 +177,6 @@ bool HttpClient::parseHeaders(bool checkTimestampOnly, uint64_t storedTimestamp)
       if (line.startsWith("PreciseSleep"))
       {
         m_sleepDuration = line.substring(14).toInt();
-
-        // Deal with program runtime compensation (if not already set) for more accurate sleep time in seconds
-        if (StateManager::getProgramRuntimeCompensationStart() == 0)
-          StateManager::setProgramRuntimeCompensationStart(millis());
-
         Serial.print("[HEADER] Sleep in seconds: ");
         Serial.println(m_sleepDuration);
       }
@@ -216,9 +251,9 @@ bool HttpClient::parseHeaders(bool checkTimestampOnly, uint64_t storedTimestamp)
   return true;
 }
 
-bool HttpClient::checkForUpdate(const String &sensorData)
+bool HttpClient::checkForUpdate(bool timestampCheck)
 {
-  if (!sendRequest(true, sensorData))
+  if (!sendRequest(timestampCheck))
     return false;
 
   if (!parseHeaders(true, StateManager::getTimestamp()))
@@ -236,7 +271,7 @@ bool HttpClient::checkForUpdate(const String &sensorData)
 
 bool HttpClient::startImageDownload()
 {
-  if (!sendRequest(false, ""))
+  if (!sendRequest(false))
     return false;
 
   if (!parseHeaders(false, 0))

--- a/src/http_client.h
+++ b/src/http_client.h
@@ -25,7 +25,7 @@ public:
   HttpClient();
 
   // Check if server has new content
-  bool checkForUpdate(const String &sensorData = "");
+  bool checkForUpdate(bool timestampCheck = true);
 
   // Start downloading image (call after checkForUpdate returns true)
   bool startImageDownload();
@@ -68,9 +68,11 @@ private:
   uint8_t m_displayRotation;
   bool m_hasRotation;
   bool m_partialRefresh;
+  String m_jsonPayload;
 
   // Internal helpers
-  bool sendRequest(bool timestampCheckOnly, const String &extraParams);
+  void buildJsonPayload();
+  bool sendRequest(bool timestampCheck);
   bool parseHeaders(bool checkTimestampOnly, uint64_t storedTimestamp);
 };
 

--- a/src/image_handler.cpp
+++ b/src/image_handler.cpp
@@ -456,9 +456,9 @@ void readImageData(HttpClient &http)
     size_t rowSize = Display::getWidth();
 
     if (streamMgr.init(rowSize))
-      Serial.println("[Image] Streaming enabled");
+      Serial.println("[IMAGE] Streaming enabled");
     else
-      Serial.println("[Image] Streaming init failed, falling back to direct mode");
+      Serial.println("[IMAGE] Streaming init failed, falling back to direct mode");
   }
 
   // Print memory stats
@@ -466,7 +466,7 @@ void readImageData(HttpClient &http)
   {
     size_t totalHeap, freeHeap, bufferUsed;
     streamMgr.getMemoryStats(totalHeap, freeHeap, bufferUsed);
-    Serial.printf("[Image] Memory - Total: %zu, Free: %zu, Buffer: %zu\n", totalHeap, freeHeap, bufferUsed);
+    Serial.printf("[IMAGE] Memory - Total: %zu, Free: %zu, Buffer: %zu\n", totalHeap, freeHeap, bufferUsed);
   }
 #endif
 

--- a/src/main.cpp
+++ b/src/main.cpp
@@ -11,6 +11,7 @@
  * Libraries:
  * EPD library: https://github.com/ZinggJM/GxEPD2
  * EPD library for 4G "Grayscale": https://github.com/ZinggJM/GxEPD2_4G
+ * ArduinoJSON: https://github.com/bblanchon/ArduinoJson
  * PNG Decoder Library: https://github.com/kikuchan/pngle
  * WiFi manager by tzapu https://github.com/tzapu/WiFiManager
  * QRCode generator: https://github.com/ricmoo/QRCode
@@ -62,35 +63,6 @@ void configModeCallback()
 // Helper Functions
 ///////////////////////////////////////////////
 
-String getSensorData()
-{
-#ifdef SENSOR
-  float temperature;
-  int humidity;
-  int pressure;
-  if (Sensor::readSensorsVal(temperature, humidity, pressure))
-  {
-    String data = "&s=" + String(Sensor::getSensorTypeStr());
-    data += "&temp=" + String(temperature) + "&hum=" + String(humidity);
-
-    switch (Sensor::getSensorType())
-    {
-      case Sensor::SensorType::BME280:
-        data += "&pres=" + String(pressure);
-        break;
-      case Sensor::SensorType::SCD4X:
-      case Sensor::SensorType::STCC4:
-        data += "&co2=" + String(pressure);
-        break;
-      default:
-        break;
-    }
-    return data;
-  }
-#endif
-  return "";
-}
-
 void initializeWiFi()
 {
   String hostname = "INK_" + Wireless::getMacAddress();
@@ -126,6 +98,9 @@ void downloadAndDisplayImage(HttpClient &httpClient)
   // Store number of pages needed to fill the buffer of the display to turn off the WiFi after last page is loaded
   uint16_t pagesToLoad = Display::getNumberOfPages();
 
+  // Start tracking download duration
+  StateManager::startDownloadTimer();
+
   do
   {
     // For paged displays, download image once per page
@@ -137,10 +112,15 @@ void downloadAndDisplayImage(HttpClient &httpClient)
     if (--pagesToLoad == 0)
     {
       httpClient.stop();
+      // End download timing before WiFi turns off
+      StateManager::endDownloadTimer();
       Wireless::turnOff();
 
       // Enable light sleep during display refresh to save power
       Display::enableLightSleepDuringRefresh(true);
+
+      // Start refresh timing after WiFi is off
+      StateManager::startRefreshTimer();
     }
   } while (Display::setToNextPage());
 
@@ -150,6 +130,9 @@ void downloadAndDisplayImage(HttpClient &httpClient)
   // Disable ePaper power
   delay(100);
   Board::setEPaperPowerOn(false);
+
+  // End refresh timing
+  StateManager::endRefreshTimer();
 
 #ifdef ES3ink
   // Show success indicator
@@ -162,9 +145,8 @@ void handleConnectedState()
   StateManager::resetFailureCount();
 
   HttpClient httpClient;
-  String sensorData = getSensorData();
 
-  if (httpClient.checkForUpdate(sensorData))
+  if (httpClient.checkForUpdate())
   {
     Serial.println("[IMAGE] Update available, downloading...");
     downloadAndDisplayImage(httpClient);
@@ -177,7 +159,7 @@ void handleConnectedState()
 
 void handleDisconnectedState()
 {
-  Serial.println("[WiFi] No Wi-Fi connection, failure count: " + String(StateManager::getFailureCount()));
+  Serial.println("[WIFI] No Wi-Fi connection, failure count: " + String(StateManager::getFailureCount()));
 
   // Calculate and set sleep duration based on failure count
   uint64_t sleepDuration = StateManager::calculateSleepDuration();
@@ -193,24 +175,22 @@ void handleDisconnectedState()
 void enterDeepSleepMode()
 {
   uint64_t sleepDuration = StateManager::getSleepDuration();
-  // Calculate compensation in milliseconds
-  unsigned long programRuntimeCompensationMs = millis() - StateManager::getProgramRuntimeCompensationStart();
 
-  // Save compensation time in milliseconds for next run (survives deep sleep)
-  StateManager::setLastCompensationTime(programRuntimeCompensationMs);
+  // Get total compensation from download + refresh durations (in milliseconds)
+  unsigned long totalCompensation = StateManager::getTotalCompensation();
 
   // Convert to seconds for sleep duration adjustment, capped to max 60 seconds
-  unsigned long programRuntimeCompensation = programRuntimeCompensationMs / 1000;
-  if (programRuntimeCompensation > 60)
-    programRuntimeCompensation = 60;
+  unsigned long compensationSeconds = totalCompensation / 1000;
+  if (compensationSeconds > 60)
+    compensationSeconds = 60;
 
-  if (programRuntimeCompensation < sleepDuration)
-    sleepDuration -= programRuntimeCompensation;
+  if (compensationSeconds < sleepDuration)
+    sleepDuration -= compensationSeconds;
 
   Serial.print("[SLEEP] Going to sleep for (seconds): ");
   Serial.print(sleepDuration);
   Serial.print(" (compensated by ");
-  Serial.print(programRuntimeCompensation);
+  Serial.print(compensationSeconds);
   Serial.println(" seconds)");
 
   Board::enterDeepSleepMode(sleepDuration);

--- a/src/sensor.cpp
+++ b/src/sensor.cpp
@@ -36,15 +36,15 @@ static bool readSTCC4(float &sen_temp, int &sen_humi, int &sen_pres);
 
 void init()
 {
-  StateManager::ResetReason reason = StateManager::getResetReason();
+  ResetReason reason = Board::getResetReason();
 
   // Reset sensor detection on power-on or hardware reset
-  if (reason == StateManager::ResetReason::POWERON || reason == StateManager::ResetReason::EXT)
+  if (reason == ResetReason::POWERON || reason == ResetReason::EXT)
   {
     Serial.println("[SENSOR] Fresh boot - resetting detection");
     detectedSensor = SensorType::NONE;
   }
-  else if (reason == StateManager::ResetReason::DEEPSLEEP)
+  else if (reason == ResetReason::DEEPSLEEP)
   {
     Serial.print("[SENSOR] Wake from deep sleep - using cached sensor: ");
     Serial.println(getSensorTypeStr());
@@ -316,6 +316,49 @@ const char *getSensorTypeStr()
       return "UNKNOWN";
   }
 }
+
+SensorData getSensorData()
+{
+  SensorData data = {nullptr, 0.0f, 0, 0, false, false, false};
+
+  float temperature;
+  int humidity;
+  int pressure = 0;
+
+  if (readSensorsVal(temperature, humidity, pressure))
+  {
+    SensorType sensorType = getSensorType();
+    data.type = getSensorTypeStr();
+    data.temperature = temperature;
+    data.humidity = humidity;
+    data.pressureOrCO2 = pressure;
+    data.isPressure = (sensorType == SensorType::BME280);
+    // Only BME280, SCD4X, and STCC4 have a third measurement
+    data.hasThirdMeasurement =
+      (sensorType == SensorType::BME280 || sensorType == SensorType::SCD4X || sensorType == SensorType::STCC4);
+    data.isValid = true;
+  }
+
+  return data;
+}
+
+} // namespace Sensor
+
+#else
+
+// Stub implementations when SENSOR is not defined
+namespace Sensor
+{
+
+void init() {}
+
+bool readSensorsVal(float &sen_temp, int &sen_humi, int &sen_pres) { return false; }
+
+SensorType getSensorType() { return SensorType::NONE; }
+
+const char *getSensorTypeStr() { return "NONE"; }
+
+SensorData getSensorData() { return {nullptr, 0.0f, 0, 0, false, false, false}; }
 
 } // namespace Sensor
 

--- a/src/sensor.h
+++ b/src/sensor.h
@@ -9,6 +9,7 @@
 // #define SENSOR
 
 #include <Arduino.h>
+#include <ArduinoJson.h>
 
 namespace Sensor
 {
@@ -27,6 +28,42 @@ bool readSensorsVal(float &sen_temp, int &sen_humi, int &sen_pres);
 
 SensorType getSensorType();
 const char *getSensorTypeStr();
+
+// Sensor data structure for passing sensor readings
+struct SensorData
+{
+  const char *type;
+  float temperature;
+  int humidity;
+  int pressureOrCO2;
+  bool isPressure;          // true = pressure (BME280), false = CO2 (SCD4X, STCC4)
+  bool hasThirdMeasurement; // true if sensor provides pressure or CO2
+  bool isValid;
+
+  // Serialize sensor data to JSON object
+  template <typename JsonArrayT>
+  void toJson(JsonArrayT &sensors) const
+  {
+    if (!isValid || type == nullptr)
+      return;
+
+    auto sensor = sensors.template add<JsonObject>();
+    sensor["type"] = type;
+    sensor["temp"] = temperature;
+    sensor["hum"] = humidity;
+
+    // Only include pressure/CO2 if the sensor provides it
+    if (hasThirdMeasurement)
+    {
+      if (isPressure)
+        sensor["pres"] = pressureOrCO2;
+      else
+        sensor["co2"] = pressureOrCO2;
+    }
+  }
+};
+
+SensorData getSensorData();
 
 } // namespace Sensor
 

--- a/src/state_manager.cpp
+++ b/src/state_manager.cpp
@@ -3,30 +3,49 @@
 // RTC persistent data (survives deep sleep)
 RTC_DATA_ATTR uint64_t rtc_timestamp = 0;
 RTC_DATA_ATTR uint8_t rtc_failureCount = 0;
-RTC_DATA_ATTR long rtc_lastCompensationTime = 0;
+RTC_DATA_ATTR unsigned long rtc_lastDownloadDuration = 0;
+RTC_DATA_ATTR unsigned long rtc_lastRefreshDuration = 0;
 
 namespace StateManager
 {
 
 uint64_t sleepDuration = DEFAULT_SLEEP_SECONDS;
-unsigned long programRuntimeCompensation = 0; // in milliseconds
+unsigned long downloadDuration = 0;
+unsigned long refreshDuration = 0;
+unsigned long downloadStartTime = 0;
+unsigned long refreshStartTime = 0;
 
 uint64_t getTimestamp() { return rtc_timestamp; }
 
 void setTimestamp(uint64_t ts) { rtc_timestamp = ts; }
 
-unsigned long getProgramRuntimeCompensationStart() { return programRuntimeCompensation; }
+void startDownloadTimer() { downloadStartTime = millis(); }
 
-void setProgramRuntimeCompensationStart(unsigned long compensation)
+void endDownloadTimer()
 {
-  // Only set once if not already set
-  if (programRuntimeCompensation == 0)
-    programRuntimeCompensation = compensation;
+  if (downloadStartTime > 0)
+  {
+    downloadDuration = millis() - downloadStartTime;
+    rtc_lastDownloadDuration = downloadDuration;
+  }
 }
 
-long getLastCompensationTime() { return rtc_lastCompensationTime; }
+void startRefreshTimer() { refreshStartTime = millis(); }
 
-void setLastCompensationTime(long time) { rtc_lastCompensationTime = time; }
+void endRefreshTimer()
+{
+  if (refreshStartTime > 0)
+  {
+    refreshDuration = millis() - refreshStartTime;
+    rtc_lastRefreshDuration = refreshDuration;
+  }
+}
+
+unsigned long getTotalCompensation() { return downloadDuration + refreshDuration; }
+
+unsigned long getLastDownloadDuration() { return rtc_lastDownloadDuration; }
+
+unsigned long getLastRefreshDuration() { return rtc_lastRefreshDuration; }
 
 uint8_t getFailureCount() { return rtc_failureCount; }
 
@@ -58,34 +77,4 @@ uint64_t calculateSleepDuration()
   return sleepDuration;
 }
 
-ResetReason getResetReason()
-{
-  esp_reset_reason_t reason = esp_reset_reason();
-
-  switch (reason)
-  {
-    case ESP_RST_POWERON:
-      return ResetReason::POWERON;
-    case ESP_RST_EXT:
-      return ResetReason::EXT;
-    case ESP_RST_SW:
-      return ResetReason::SW;
-    case ESP_RST_PANIC:
-      return ResetReason::PANIC;
-    case ESP_RST_INT_WDT:
-      return ResetReason::INT_WDT;
-    case ESP_RST_TASK_WDT:
-      return ResetReason::TASK_WDT;
-    case ESP_RST_WDT:
-      return ResetReason::WDT;
-    case ESP_RST_DEEPSLEEP:
-      return ResetReason::DEEPSLEEP;
-    case ESP_RST_BROWNOUT:
-      return ResetReason::BROWNOUT;
-    case ESP_RST_SDIO:
-      return ResetReason::SDIO;
-    default:
-      return ResetReason::UNKNOWN;
-  }
-}
 } // namespace StateManager

--- a/src/state_manager.h
+++ b/src/state_manager.h
@@ -6,24 +6,6 @@
 namespace StateManager
 {
 
-enum class ResetReason : uint8_t
-{
-  UNKNOWN = 0,
-  POWERON = 1,   // Power-on reset or flash upload
-  EXT = 2,       // External reset (reset button)
-  SW = 3,        // Software reset
-  PANIC = 4,     // Software panic/exception
-  INT_WDT = 5,   // Interrupt watchdog
-  TASK_WDT = 6,  // Task watchdog
-  WDT = 7,       // Other watchdog
-  DEEPSLEEP = 8, // Wake from deep sleep
-  BROWNOUT = 9,  // Brownout reset
-  SDIO = 10      // Reset over SDIO
-};
-
-// Reset reason
-ResetReason getResetReason();
-
 // Timestamp management
 uint64_t getTimestamp();
 void setTimestamp(uint64_t ts);
@@ -38,13 +20,14 @@ uint64_t getSleepDuration();
 void setSleepDuration(uint64_t seconds);
 uint64_t calculateSleepDuration();
 
-// Program runtime compensation
-unsigned long getProgramRuntimeCompensationStart();
-void setProgramRuntimeCompensationStart(unsigned long compensation);
-
-// Last runtime compensation time (in RTC memory to send as part of a report to server next run)
-long getLastCompensationTime();
-void setLastCompensationTime(long time);
+// Download/display refresh duration tracking
+void startDownloadTimer();
+void endDownloadTimer();
+void startRefreshTimer();
+void endRefreshTimer();
+unsigned long getTotalCompensation();
+unsigned long getLastDownloadDuration();
+unsigned long getLastRefreshDuration();
 
 // Default sleep time
 static const uint64_t DEFAULT_SLEEP_SECONDS = 120;

--- a/src/streaming_handler.cpp
+++ b/src/streaming_handler.cpp
@@ -22,19 +22,19 @@ bool RowStreamBuffer::init(size_t rowSizeBytes, size_t rowCount)
 {
   if (m_initialized)
   {
-    Serial.println("[Streaming] RowBuffer already initialized");
+    Serial.println("[STREAM] RowBuffer already initialized");
     return true;
   }
 
   if (rowSizeBytes == 0 || rowSizeBytes > MAX_ROW_SIZE)
   {
-    Serial.printf("[Streaming] Invalid row size: %zu (max: %zu)\n", rowSizeBytes, MAX_ROW_SIZE);
+    Serial.printf("[STREAM] Invalid row size: %zu (max: %zu)\n", rowSizeBytes, MAX_ROW_SIZE);
     return false;
   }
 
   if (rowCount == 0)
   {
-    Serial.println("[Streaming] Invalid row count: 0");
+    Serial.println("[STREAM] Invalid row count: 0");
     return false;
   }
 
@@ -44,7 +44,7 @@ bool RowStreamBuffer::init(size_t rowSizeBytes, size_t rowCount)
   size_t freeHeap = Utils::getFreeHeap();
   if (freeHeap < totalSize * 2)
   {
-    Serial.printf("[Streaming] Insufficient heap: %zu bytes free, need %zu for row buffer\n", freeHeap, totalSize * 2);
+    Serial.printf("[STREAM] Insufficient heap: %zu bytes free, need %zu for row buffer\n", freeHeap, totalSize * 2);
     return false;
   }
 
@@ -55,13 +55,13 @@ bool RowStreamBuffer::init(size_t rowSizeBytes, size_t rowCount)
     m_rowSize = rowSizeBytes;
     m_rowCount = rowCount;
     m_initialized = true;
-    Serial.printf("[Streaming] Row buffer initialized: %zu bytes/row × %zu rows = %zu bytes total\n", rowSizeBytes,
+    Serial.printf("[STREAM] Row buffer initialized: %zu bytes/row × %zu rows = %zu bytes total\n", rowSizeBytes,
                   rowCount, totalSize);
     return true;
   }
   catch (const std::bad_alloc &e)
   {
-    Serial.printf("[Streaming] Row buffer allocation failed: %s\n", e.what());
+    Serial.printf("[STREAM] Row buffer allocation failed: %s\n", e.what());
     return false;
   }
 }
@@ -73,7 +73,7 @@ size_t RowStreamBuffer::writeRow(size_t rowIndex, const uint8_t *data, size_t le
 
   if (rowIndex >= m_rowCount)
   {
-    Serial.printf("[Streaming] Invalid row index: %zu (max: %zu)\n", rowIndex, m_rowCount - 1);
+    Serial.printf("[STREAM] Invalid row index: %zu (max: %zu)\n", rowIndex, m_rowCount - 1);
     return 0;
   }
 
@@ -125,18 +125,18 @@ bool StreamingManager::init(size_t rowSizeBytes, size_t rowCount)
 {
   if (m_enabled)
   {
-    Serial.println("[Streaming] Manager already enabled");
+    Serial.println("[STREAM] Manager already enabled");
     return true;
   }
 
   if (!m_buffer.init(rowSizeBytes, rowCount))
   {
-    Serial.println("[Streaming] Failed to initialize row buffer");
+    Serial.println("[STREAM] Failed to initialize row buffer");
     return false;
   }
 
   m_enabled = true;
-  Serial.println("[Streaming] Manager initialized successfully");
+  Serial.println("[STREAM] Manager initialized successfully");
   return true;
 }
 
@@ -153,7 +153,7 @@ void StreamingManager::cleanup()
   {
     m_buffer.clear();
     m_enabled = false;
-    Serial.println("[Streaming] Manager cleanup complete");
+    Serial.println("[STREAM] Manager cleanup complete");
   }
 }
 

--- a/src/utils.cpp
+++ b/src/utils.cpp
@@ -20,7 +20,7 @@ size_t getLargestFreeBlock() { return ESP.getMaxAllocHeap(); }
 
 void printMemoryStats()
 {
-  Serial.println("[Memory Stats]");
+  Serial.println("[MEMORY]");
   Serial.printf("  Total Heap:  %zu bytes\n", getTotalHeap());
   Serial.printf("  Free Heap:   %zu bytes\n", getFreeHeap());
   Serial.printf("  Largest Block: %zu bytes\n", getLargestFreeBlock());
@@ -36,13 +36,13 @@ void initializeAPIKey()
     // Generate 8-digit PIN without leading zeros (range: 10000000 - 99999999)
     rtc_cachedPIN = (esp_random() % 90000000) + 10000000;
     prefs.putULong("apikey", rtc_cachedPIN);
-    Serial.print("Generated new device API key: ");
+    Serial.print("[APIKEY] Generated new device API key: ");
     Serial.println(rtc_cachedPIN);
   }
   else
   {
     rtc_cachedPIN = prefs.getULong("apikey");
-    Serial.print("Device API key: ");
+    Serial.print("[APIKEY] Device API key: ");
     Serial.println(rtc_cachedPIN);
   }
 

--- a/src/wireless.cpp
+++ b/src/wireless.cpp
@@ -21,7 +21,7 @@ void init(const String &hostname, const String &password, void (*callback)())
 {
   // Connecting to WiFi
   Serial.println();
-  Serial.print("[WiFi] Connecting... ");
+  Serial.print("[WIFI] Connecting... ");
   WiFi.mode(WIFI_STA);
   wm.setWiFiAutoReconnect(true);
   wm.setConnectRetries(5);
@@ -42,7 +42,7 @@ void init(const String &hostname, const String &password, void (*callback)())
 String getSSID()
 {
   String in = WiFi.SSID();
-  Serial.println("[WiFi] SSID: " + in);
+  Serial.println("[WIFI] SSID: " + in);
   if (in.length() == 0)
     return in;
 
@@ -74,7 +74,7 @@ String getSSID()
 int8_t getStrength()
 {
   int8_t rssi = WiFi.RSSI();
-  Serial.println("[WiFi] Strength: " + String(rssi) + " dB");
+  Serial.println("[WIFI] Strength: " + String(rssi) + " dB");
   return rssi;
 }
 
@@ -84,6 +84,8 @@ String getSoftAPSSID() { return WiFi.softAPSSID(); }
 
 String getSoftAPIP() { return WiFi.softAPIP().toString(); }
 
+String getIPAddress() { return WiFi.localIP().toString(); }
+
 bool isConnected() { return WiFi.status() == WL_CONNECTED; }
 
 void turnOff()
@@ -91,7 +93,7 @@ void turnOff()
   WiFi.disconnect(true);
   WiFi.mode(WIFI_OFF);
   delay(20);
-  Serial.println("[WiFi] Turned off");
+  Serial.println("[WIFI] Turned off");
 }
 
 void resetCredentialsAndReboot()
@@ -100,7 +102,7 @@ void resetCredentialsAndReboot()
   turnOff();
 
   // Reset WiFi settings (erase stored credentials)
-  Serial.println("[WiFi] Erasing stored credentials...");
+  Serial.println("[WIFI] Erasing stored credentials...");
   wm.resetSettings();
 
   // Restart ESP to start configuration portal

--- a/src/wireless.h
+++ b/src/wireless.h
@@ -12,6 +12,7 @@ String getMacAddress();
 
 String getSoftAPSSID();
 String getSoftAPIP();
+String getIPAddress();
 
 bool isConnected();
 void turnOff();

